src/util/antlr/ANTLRErrorHandling.cpp

```diff
@@ -2,38 +2,76 @@
+}
+
+// _____________________________________________________________________________
+ExceptionMetadata generateMetadata(antlr4::Lexer* lexer, size_t line,
+                                   size_t charPositionInLine) {
+  return {lexer->getInputStream()->toString(), lexer->tokenStartCharIndex,
+          lexer->tokenStartCharIndex, line, charPositionInLine};
+}
 
 // _____________________________________________________________________________
 ExceptionMetadata generateMetadata(antlr4::Recognizer* recognizer,
-                                   antlr4::Token* offendingToken,
-                                   [[maybe_unused]] size_t line,
-                                   [[maybe_unused]] size_t charPositionInLine) {
-  return {((antlr4::CommonTokenStream*)recognizer->getInputStream())
-              ->getTokenSource()
-              ->getInputStream()
-              ->toString(),
-          offendingToken->getStartIndex(), offendingToken->getStopIndex()};
+                                   antlr4::Token* offendingToken, size_t line,
+                                   size_t charPositionInLine) {
+  // The `antlr4::Recognizer` interface does not provide enough information to
+  // obtain a copy of the input string. Do this switching here to obtain that.
+  // As of ANTLRv4 `antlr4::Parser` and `antlr4::Lexer` are the only subclasses
+  // of `antlr4::Recognizer`.
+  if (auto* parser = dynamic_cast<antlr4::Parser*>(recognizer)) {
+    AD_CHECK(offendingToken != nullptr);
+    return generateMetadata(parser, offendingToken, line, charPositionInLine);
+  } else if (auto* lexer = dynamic_cast<antlr4::Lexer*>(recognizer)) {
+    // If the recognizer is a Lexer this means that the error was a Lexer error.
+    // In that case the tokens are not yet available and `offendingToken` is
+    // NULL.
+    return generateMetadata(lexer, line, charPositionInLine);
+  } else {
+    AD_FAIL();  // Should be unreachable.
+  }
 }
 
 // _____________________________________________________________________________
 ExceptionMetadata generateMetadata(antlr4::ParserRuleContext* ctx) {
-  return {ctx->getStart()->getInputStream()->toString(),
-          ctx->getStart()->getStartIndex(), ctx->getStop()->getStopIndex()};
+  auto start = ctx->getStart();
+  return {start->getInputStream()->toString(), ctx->getStart()->getStartIndex(),
+          ctx->getStop()->getStopIndex(), start->getLine(),
+          start->getCharPositionInLine()};
+}
+
+std::string generateExceptionMessage(antlr4::Token* offendingSymbol,
+                                     size_t line, size_t charPositionInLine,
+                                     const std::string& msg) {
+  return (offendingSymbol)
+             ? absl::StrCat("Token \"", offendingSymbol->getText(),
+                            "\" at line ", line, ":", charPositionInLine, " ",
+                            msg)
+             : absl::StrCat("line ", line, ":", charPositionInLine, " ", msg);
```

Comment:
I am not sure, whether the line and position should ALWAYS be part of the message. But we can first merge this as is, and then get some experience with displaying parse errors in the UI. But maybe we want the ParseException to have an interface `errorMessageIncludingPositionalInfo` and `errorMessageWithoutPositionalInfo`.

Reply:
(the latter one would just be the `msg`.